### PR TITLE
Add SameSite to Cookies

### PR DESCRIPTION
**Pull request by cfabianski**
### Summary

Add `SameSite` to the cookies to reinforce the security. 
Define `Lax` as a first step.
 `Strict` is also supported but is too strict as its name suggests 😸

### Other Information

[`SameSite`](https://tools.ietf.org/html/draft-west-first-party-cookies-07) is the new recommended way to keep your website secure. It would make `CSRF` even more secure.

Some links:

- https://scotthelme.co.uk/csrf-is-dead/
- https://www.sjoerdlangkemper.nl/2016/04/14/preventing-csrf-with-samesite-cookie-attribute/
- https://chloe.re/2016/04/13/goodbye-csrf-samesite-to-the-rescue/
- https://tools.ietf.org/html/draft-west-first-party-cookies-07

### CODE REVIEW

**Comment by pixeltrix** [comment]
@cfabianski is there any risk of breaking an app when upgrading from an older version if we now enable `SameSite` by default. Also I don't see a test for making sure that it can be disabled.

Generally though 👍

**Comment by cfabianski** [comment]
We could have a global setting to make it backward compatible. Though, I thought `Lax` would be safe enough to start with.

File: railties/lib/rails/generators/rails/app/templates/config/initializers/cookies_same_site_protection.rb
Comment (AndyStabler) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/cookies_same_site_protection.rb]: There's a random `a` here

**Review by AndyStabler — CHANGES_REQUESTED** [review]
looks good, just noticed a teeny tiny typo

File: railties/lib/rails/generators/rails/app/templates/config/initializers/cookies_same_site_protection.rb
Comment (cfabianski) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/cookies_same_site_protection.rb]: Updated. Thanks @AndyStabler

**Review by AndyStabler — APPROVED** [review]
🎉

**Comment by cfabianski** [comment]
@pixeltrix does it look better now?

**Comment by cfabianski** [comment]
@maclover7 Do you have any thoughts about it?

File: actionpack/lib/action_dispatch/middleware/cookies.rb
Comment (maclover7) [inline actionpack/lib/action_dispatch/middleware/cookies.rb]: I'm probably just missing it, but where is this method used? I didn't see any calls in the PR.

**Review by maclover7 — CHANGES_REQUESTED** [review]
Coming along very nicely. Two quick things:

- Can you also add an entry to Action Pack's CHANGELOG.md about this?
- Can you add a test for what happens to the cookies when the config value is `:no_protection`?

File: actionpack/lib/action_dispatch/middleware/cookies.rb
Comment (cfabianski) [inline actionpack/lib/action_dispatch/middleware/cookies.rb]: It's not but I followed the pattern which is to have access to this as part of the cookie. Want me to remove it?

Comment (maclover7) [inline actionpack/lib/action_dispatch/middleware/cookies.rb]: If you're not calling it anywhere directly, then I'd say to not add the method right now. If it's needed in the future, it's easy to add :)

**Comment by cfabianski** [comment]
@maclover7 Done

File: actionpack/lib/action_dispatch/middleware/cookies.rb
Comment (cfabianski) [inline actionpack/lib/action_dispatch/middleware/cookies.rb]: Here we go :)

**Review by maclover7 — APPROVED** [review]
👍 looks good to me, but Security Team folks will probably want to review this 😬

**Comment by cfabianski** [comment]
@maclover7 great thanks. 
Do I need to contact the security team or will they just review it when available?

**Comment by maclover7** [comment]
@cfabianski They'll review it when they get the chance

**Comment by cfabianski** [comment]
@maclover7 do you have any update on that one?
I can resolve the conflicts but it would be nice to get an idea on whether or not this is something we might want.

**Comment by swrobel** [comment]
Maybe a good feature for Rails 6?

**Comment by cfabianski** [comment]
@swrobel I hope so! 😺

**Comment by cfabianski** [comment]
@swrobel fixed the conflicts just in case somebody wants to make it part of Rails 6 :) /cc @maclover7

**Comment by cfabianski** [comment]
I'll have a look at the failures and I'll get them fixed in the meantime

**Review by gmcgibbon — COMMENTED** [review]
`SameSite` is still absent in Rails 6 (2d3a28b949f2e8a0b755afdb52cbcd55e5ac0d98) cookie headers. I think it should be added! 👍 If you're still interested in getting this merged, please rebase and squash your commits.

**Comment by cfabianski** [comment]
Hey @gmcgibbon, thank you for your review. I managed to implement the changes you asked for.
I've also squashed the commits. Let me know if there is anything I can do :)

**Review by gmcgibbon — COMMENTED** [review]
👍 Code and tests look great but I'm still a little hesitant on the changelog entry.

File: actionpack/CHANGELOG.md
Comment (cfabianski) [reply actionpack/CHANGELOG.md]: I missed the update here sorry. I've updated it as well now (squashed)

File: actionpack/CHANGELOG.md
Comment (gmcgibbon) [inline actionpack/CHANGELOG.md]: For this sentence, I would say
```
By switching from `:no_protection` to `:lax` in `config/initializers/cookies.rb`,
cookies won't be sent by browsers in cross-site POST requests.
`:strict` will disable cookies from being sent in both cross-site GET and POST requests.
```
to provide a little more detail on this feature.

**Review by gmcgibbon — COMMENTED** [review]
One last comment regarding the changelog and then this should be good to go! 😄

**Comment by cfabianski** [comment]
@gmcgibbon Thank you. I've updated it!

**Review by gmcgibbon — APPROVED** [review]
💯

**Comment by cfabianski** [comment]
@gmcgibbon let's hope we can get it merged now :)

**Comment by dritir** [comment]
@cfabianski Does this change will give the option to enable SameSite on session cookies aswell?

File: actionpack/lib/action_dispatch/middleware/cookies.rb
Comment (jeremy) [inline actionpack/lib/action_dispatch/middleware/cookies.rb]: If someone makes a typo in the config value, the SameSite option will be ignored. [Rack validates](https://github.com/rack/rack/pull/1051) allowed `:same_site` values. Can we rely on that?

File: actionpack/CHANGELOG.md
Comment (jeremy) [inline actionpack/CHANGELOG.md]: Setting `same_site: …` is already possible with cookies since it's supported in Rack. Let's clarify that this makes it a default setting that applies to all cookies.

File: activestorage/test/dummy/config/initializers/cookies.rb
Comment (jeremy) [inline activestorage/test/dummy/config/initializers/cookies.rb]: This config is feeling like a special case for one type of default cookie option.

How would it compare with offering config for cookie defaults, and setting `same_site: nil` as the default?

**Comment by ignisf** [comment]
Hello,
Would the transition be seamless or would one have through a migration period?

**Comment by sikachu** [comment]
Hello~

At this point, it might be late or rush to put it in 6.0. However, I agree that this is a good idea, so I still would like to push it at least for 6.1.

@cfabianski, will you have some time to respond to @jeremy's feedback above? Thank you.

File: railties/lib/rails/generators/rails/app/templates/config/initializers/cookies.rb.tt
Comment (sikachu) [inline railties/lib/rails/generators/rails/app/templates/config/initializers/cookies.rb.tt]: I think this config should instead be live in environment configuration or application configuration.

We can, say, have the default set to `:no_protection` by default, but then set this to `:lax` in production.rb template, and in the new framework default.

**Comment by cfabianski** [comment]
> @cfabianski, will you have some time to respond to @jeremy's feedback above? Thank you.

I'll work on that hopefully soon. Thank you @sikachu and @jeremy

**Comment by cfabianski** [comment]
> I'll work on that hopefully soon. Thank you @sikachu and @jeremy

@sikachu and @jeremy I've addressed your comments. Please have another look and let me know if that's better :)

**Comment by LeoJweda** [comment]
@cfabianski: There are conflicts and a 3 outstanding comments from @jeremy. Is this PR still actively being worked on?

**Comment by ignisf** [comment]
I was under the impression that @jeremy's comments were addressed?

**Comment by LeoJweda** [comment]
My apologies. I see now that the last commit did address those comments, they just weren't marked as outdated.

**Comment by ignisf** [comment]
@sikachu, do you think there is still chance this could ship with 6.1?

**Comment by gaffneyc** [comment]
With `SameSite=Lax` becoming the default should an option for `:none` be added to opt out completely?

Also, specifying `:no_protection` seems to triggers the browser's default behavior. Older browsers this would be the equivalent of 'None' but newer browsers will be 'Lax' protection. Because of that it would probably be better to either have it set `None` or rename the option.

https://scotthelme.co.uk/csrf-is-really-dead/

**Comment by grzuy** [comment]
> With `SameSite=Lax` becoming the default should an option for `:none` be added to opt out completely?
> 
> Also, specifying `:no_protection` seems to triggers the browser's default behavior. Older browsers this would be the equivalent of 'None' but newer browsers will be 'Lax' protection. Because of that it would probably be better to either have it set `None` or rename the option.
> 
> https://scotthelme.co.uk/csrf-is-really-dead/

@gaffneyc Agree.

Something like
```rb
# Possible values
:none
:browser_default # (or false / nil)
:lax
:strict
```
?

**Comment by cfabianski** [comment]
I'd like to know if there is any interest in getting this PR merged from a Rails Core Member first. This PR is more than 2 years old :(

**Comment by aphexddb** [comment]
Chrome is now displaying errors as follows:

> A cookie associated with a cross-site resource at http://yourwebsitehere.com/ was set without the `SameSite` attribute. A future release of Chrome will only deliver cookies with cross-site requests if they are set with `SameSite=None` and `Secure`.

According to the [Chrome release plan](https://www.chromium.org/updates/same-site), starting February 4, 2020 when Chrome 80 Stable is released the SameSite-by-default and SameSite=None-requires-Secure become default behavior for all users on Stable.

**Comment by gaffneyc** [comment]
@martin-svk [Per the Rails maintenance policy](https://rubyonrails.org/maintenance/) this would be part of Rails 6.1 (at the earliest) and not back ported to older versions of Rails. Rails 4 is currently End of Life and won't be receiving updates or security fixes.

File: actionpack/lib/action_dispatch/middleware/cookies.rb
Comment (mtomov) [reply actionpack/lib/action_dispatch/middleware/cookies.rb]: I believe there's also the value of `None`

See:

https://tools.ietf.org/html/draft-ietf-httpbis-rfc6265bis-03

```
samesite-value    = "Strict" / "Lax" / "None"
```

https://www.chromestatus.com/feature/5088147346030592
https://www.chromestatus.com/feature/5633521622188032 - `None` cookies have to be secure

Comment (HarrisonB) [reply actionpack/lib/action_dispatch/middleware/cookies.rb]: Note [this is necessary for cross-domain iframe and AJAX requests](https://makandracards.com/makandra/71018-samesite-cookies#what-this-means-for-web-development) that rely on cookies to be sent.

Comment (HarrisonB) [reply actionpack/lib/action_dispatch/middleware/cookies.rb]: Don't know if you'd wanna note that [it also is incompatible with a whole host of clients](https://www.chromium.org/updates/same-site/incompatible-clients).

**Comment by Pezmc** [comment]
Chrome for Desktop Beta 79 now has the `same-site-by-default-cookies` flag enabled by default.

More information at https://www.chromestatus.com/feature/5088147346030592 and https://www.chromestatus.com/feature/5633521622188032.

**Review by sikachu — APPROVED** [review]
@cfabianski sorry, I was busy and overlooked this PR.

Would you mind rebase this PR again and force push to your branch so we can merge this in?

**Comment by cfabianski** [comment]
@sikachu that's now rebased! Thank you.

**Comment by mtomov** [comment]
Thanks for updating @cfabianski . Any chance you'd be able to add the `SameSite=None` option as well? It's needed for cross-site requests, although with care as indicated in the comments above by @HarrisonB : https://github.com/rails/rails/pull/28297#discussion_r341815043

image

**Comment by kaspth** [comment]
Thanks for sticking with this @cfabianski ❤️ As a new chef on this thing, I ended with some different opinions on where to take this, but I didn't want to push you through another review. I made these changes:

- Upgrade the config to a new framework default and only enable it for newly generated apps.
- Since appending `SameSite=None` isn't supported by a released Rack version yet, I swapped it to `false` and opting out of adding the header. Opened this PR for the eventual follow up https://github.com/rails/rails/pull/37974
- Note: I wasn't comfortable adding the SameSite header yet because of signed/encrypted cookies and being suspicious of whether they'd start failing verification/decryption. Generally with cookies I'd prefer to set a new config once people have upgraded and verified it's stable.
- Back out of the cookies_serializer.rb -> cookies.rb rename, not needed anymore.
- Added an immediate fixup commit here c26b65382e 😅

Thanks again for pushing for this! Let me know if there's anything I've missed or misunderstood about the SameSite option. Maybe particularly around with Chrome is going to do with the missing header? Perhaps we'd like some workaround for Rails 6.0.

**Comment by kaspth** [comment]
Another thought I had, usually it takes a lot of code and back and forth to add these new headers.

Perhaps it would be nice to have a cookie config akin to the app level `force_ssl`, something like `config.action_dispatch.cookies_headers = { "SameSite": "Lax" }` then Rails 6.0 apps could upgrade their protection already. Though it's a little wild westy in that there'd be no verification of the passed hash.

**Comment by remofritzsche** [comment]
@kaspth Thanks for your work. Shouldn't this issue still be open though? It's not merged yet, right?

**Comment by kaspth** [comment]
@remofritzsche not sure what you mean? I merged it in cd1aeda unless I'm missing something.

**Comment by saragee3** [comment]
following up on this... does this need to be reopened?

**Comment by kaspth** [comment]
Why? It was merged in the commit I referenced, no?

**Comment by KelseyDH** [comment]
> @martin-svk [Per the Rails maintenance policy](https://rubyonrails.org/maintenance/) this would be part of Rails 6.1 (at the earliest) and not back ported to older versions of Rails. Rails 4 is currently End of Life and won't be receiving updates or security fixes.

Given this is set to break all Rails applications using Chrome (~80% of browser users) in 14 days... backporting this to Rails 5.x as a patch ASAP should be a high, high priority.  Two years ago Google warned about these changes and it's disappointing Rails took this long to add this configuration to the `cookies` options hash.

**Comment by rafaelfranca** [comment]
We will backport to 6.0 as a bug fix, but I don't know this warrants a backport to a security only release like 5.2. Rails 6.0 was released 6 months ago, and upgrading applications could be high, high priority if that problem is so important.

> Two years ago Google warned about these changes and it's disappointing Rails took this long to add this configuration to the `cookies` options hash.

It depends on the definition of warning. I got no email from Google telling us they would unilaterally make a change that would break a lot of things in the internet, even old versions of their own browser if some app, I don't know decide to send `SameSite=None` by default to every cookie.

**Comment by KelseyDH** [comment]
At this point I'm looking to write a Rack middleware to hack in `SameSite=None;` into all our cross-site cookies on our Rails 5.x app.  [This answer suggests something to the effect of](https://stackoverflow.com/a/3295243):

```ruby
# lib/same_site_cookie_middleware
class SameSiteCookieMiddleware
  def initialize(app)
    @app = app
  end

  def call(env)
    status, headers, body = @app.call(env)
    # confusingly, response takes its args in a different order
    # than rack requires them to be passed on
    # I know it's because most likely you'll modify the body, 
    # and the defaults are fine for the others. But, it still bothers me.

    response = Rack::Response.new body, status, headers

    response.set_cookie("foo", {:value => "bar", :path => "/", :expires => 1.year.from_now, same_site: :none})
    response.finish # finish writes out the response in the expected format.
  end
end
```

``ruby
# application.rb
require 'same_site_cookie_middleware`
config.middleware.insert_after(SameSiteCookieMiddleware, ActionDispatch::Cookies)
````

But without access to a `same_site` cookie hash attribute I'm unsure about how to bulk append `SameSite=None;` to every cookie?

**Comment by HarrisonB** [comment]
@KelseyDH This is similar to what we had to do. We also had to evaluate whether or not to actually mutate the cookie based on the useragent using the logic [suggested by the Chromium team](https://www.chromium.org/updates/same-site/incompatible-clients) as it breaks old clients (as discussed above).

As for the actual cookie editing, we did a manual string edit using regexes.

> It depends on the definition of warning. I got no email from Google telling us they would unilaterally make a change that would break a lot of things in the internet, even old versions of their own browser if some app, I don't know decide to send `SameSite=None` by default to every cookie.

As much as I hate to say it, this seems like something Rails might need to work around. Perhaps this change will get delayed, but my guess is that it will get through eventually...even Rails 6.1 will have to reckon with the choice of breaking some group of browsers or doing some useragent-specific responding.

**Comment by KelseyDH** [comment]
@HarrisonB > @KelseyDH This is similar to what we had to do. We also had to evaluate whether or not to actually mutate the cookie based on the useragent using the logic [suggested by the Chromium team](https://www.chromium.org/updates/same-site/incompatible-clients) as it breaks old clients (as discussed above).

Thanks for sharing that!  Fortunately it appears about [80% of clients are on `SameSite` compatible browsers](https://caniuse.com/#feat=same-site-cookie-attribute).

> As much as I hate to say it, this seems like something Rails might need to work around. Perhaps this change will get delayed, but my guess is that it will get through eventually...even Rails 6.1 will have to reckon with the choice of breaking some group of browsers or doing some useragent-specific responding.

Yeah Rails will probably need to begin intelligently parsing `User-Agent` headers for compatibility.

> As for the actual cookie editing, we did a manual string edit using regexes.

I would love to see this sample code if you have it. 🙏   After February 4 there are going to be a lot of Rails developers scrambling to find workarounds.

**Comment by HarrisonB** [comment]
@KelseyDH unfortunately I cannot share the code at this time.

**Comment by lukad** [comment]
@KelseyDH I can share our code. We're manipulating the `Set-Cookie` header string manually as well.

```ruby
class SameSite
  def initialize(app)
    @app = app
  end

  def call(env)
    status, headers, response = @app.call(env)

    headers['Set-Cookie'] = headers['Set-Cookie'].split("\n").map do |cookie|
      next if cookie.blank?
      next if cookie =~ /;\sSameSite/

      cookie + '; SameSite=None' + '; Secure'
    end.join("\n")

    [status, headers, response]
  end
end
```

**Comment by p8** [comment]
There's also: https://github.com/pschinis/rails_same_site_cookie

**Comment by remofritzsche** [comment]
> Given this is set to break all Rails applications using Chrome (~80% of browser users) in 14 days... backporting this to Rails 5.x as a patch ASAP should be a high, high priority. Two years ago Google warned about these changes and it's disappointing Rails took this long to add this configuration to the `cookies` options hash.

@KelseyDH: You got me startled here for a second ;) But I got [the announcement](https://blog.chromium.org/2019/10/developers-get-ready-for-new.html) right, Chrome 80 is not disposing all cookies without the `SameSite` attribute, but just treats them as `SameSite=Lax` automatically. And, additionally, it won't allow `SameSite=None` cookies anymore that are not flagged as `Secure`. So I think all Rails applications that do not rely on Cross-Site cookies should still work without any changes. Is that right?

**Comment by walkersumida** [comment]
@p8 It works on Rails 5.

**Comment by vfonic** [comment]
@rafaelfranca 
> It depends on the definition of warning. I got no email from Google telling us they would unilaterally make a change that would break a lot of things in the internet, even old versions of their own browser if some app, I don't know decide to send SameSite=None by default to every cookie.

I understand your point of view. What everyone here wants is a working internet. The fact that Google acts the way they do, doesn't mean we need to retaliate in the same way. We've been writing fixes for IE for years, not because we loved the browser, but because we wanted our users to be able to use our webapps.

Chrome is the most used browser in the world. It would be great that one of the most popular web frameworks had a full support for it.

**Comment by santib** [comment]
Question: do the changes merged mean that new Rails apps won't support older browsers? For example here it says Chrome 51 to Chrome 66 and some others will reject cookies with `SameSite=None` https://www.chromium.org/updates/same-site/incompatible-clients .

**Comment by wlipa** [comment]
The release of this in Chrome has been pushed back a little.  "The SameSite-by-default and SameSite=None-requires-Secure behaviors will begin rolling out to Chrome 80 Stable for an initial limited population starting the week of February 17, 2020, excluding the US President’s Day holiday on Monday."

But apps that need cookies to work in a cross-origin manner (like when embedded in an iframe) will need some fix soon.

https://www.chromium.org/updates/same-site

**Comment by szpasztor** [comment]
SameSite enforcements are being rolled out as of Mar 2. Any updates on this ticket? This is starting to cause breaks in some production apps.

**Comment by Fjan** [comment]
It turns out that old Safari browser are [broken](https://www.chromestatus.com/feature/5088147346030592) with SameSite:None if you rely on cross-site cookies. For the convenience of other people needing to fix this, this monkey patch wil set SameSite:None if it sees a new Chrome browser to get some immediate relieve. Just drop in config/initialisers, should work on Rails 5 & 6, but you do need the latest version of the rack gem.

```
module ActionDispatch
  class Cookies
    class CookieJar #:nodoc:
      def handle_options(options) #:nodoc:

        # Monkey patch to get app to work on Chrome 80 work inside iframes again
        if request.ssl? && request.user_agent&.include?('Chrome/8')
          options[:same_site] ||= "None";options[:secure] = true
        end

        if options[:expires].respond_to?(:from_now)
          options[:expires] = options[:expires].from_now
        end

        options[:path] ||= "/"

        if options[:domain] == :all || options[:domain] == "all"
          # If there is a provided tld length then we use it otherwise default domain regexp.
          domain_regexp = options[:tld_length] ? /([^.]+\.?){#{options[:tld_length]}}$/ : DOMAIN_REGEXP

          # If host is not ip and matches domain regexp.
          # (ip confirms to domain regexp so we explicitly check for ip)
          options[:domain] = if (request.host !~ /^[\d.]+$/) && (request.host =~ domain_regexp)
            ".#{$&}"
          end
        elsif options[:domain].is_a? Array
          # If host matches one of the supplied domains without a dot in front of it.
          options[:domain] = options[:domain].find { |domain| request.host.include? domain.sub(/^\./, "") }
        end
      end
    end
  end
end
```

**Comment by yboulkaid** [comment]
For anyone running into this, this is already possible today:

```ruby
Rails.application.config.session_store(:cookie_store, same_site: :none)
```

Note that if you want to use `none` as an option to `same_site`, you'll need to be on a `rack` version that's higher than `2.1.0`, see https://github.com/rack/rack/commit/c859bbf7b53cb59df1837612a8c330dfb4147392.

This is because rails' `CookieStore` extends `Rack::Session::Abstract::Persisted`, see https://github.com/rails/rails/blob/f33d52c95217212cbacc8d5e44b5a8e3cdc6f5b3/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb.

**Comment by Fjan** [comment]
@yboulkaid That will set it on all browsers, so that's fine but only if you do need to worry about old Safari clients.

**Comment by blackwellsmith** [comment]
How do you implement SameSite attribute for cookies in rails? Ok so I have searched and searched. May explanations on what is happening with SameSite. I can't find anything that shows me how or where to do that. It's like I am completely missing something. Can someone share some insight on how to set this attribute in Rails? I need something concise not vague. Thanks!

**Comment by blackwellsmith** [comment]
ok so an update I have SameSite=None but it need to be secure as well any ideas on how todo that?

**Comment by yeonhoyoon** [comment]
@blackwellsmith you can add `secure: true` option.

**Comment by blackwellsmith** [comment]
I tried that it didn’t work. I am currently trying it with the rails same site cookie gem and I can see the cookie. I am still trouble shooting some issues.

**Comment by shawndeprey** [comment]
@Fjan Solution above worked beautifully as a short term patch given browsers currently sparsely support this config.

It would be nice to have this type of middleware easily configurable in the Rails base config somewhere; for API only applications this new cookie issue is becoming increasingly prevalent.

**Comment by Fjan** [comment]
@shawndeprey Note that the latest Firefox also drops cookies inside iframes unless they are secure and same-site. It's safe to set this on older Firefox browsers, I changed the test in the monkey patch to:
`if request.ssl? && request.user_agent =~ /Chrome\/8|Firefox/`

**Comment by shawndeprey** [comment]
@Fjan that's good to know and thanks for the updated config. This is super helpful. :clap:

**Comment by earthctzn** [comment]
@yboulkaid I tried this on Rails 6.0.3.2 with Rack 2.2.3 and it had no effect on my `_session_id` cookie.  I only got it to work using the https://github.com/pschinis/rails_same_site_cookie that @p8 mentioned above- thanks for that btw! Here is what I tried, what did I miss? Thanks for any responses. 

In config/initializers/session_store.rb
```Rails.application.config.session_store :cookie_store, key: '_MY_APP', same_site: :none, secure: true```

> For anyone running into this, this is already possible today for session cookies:
> 
> ```ruby
> Rails.application.config.session_store(:cookie_store, same_site: :none)
> ```
> 
> Note that if you want to use `none` as an option to `same_site`, you'll need to be on a `rack` version that's higher than `2.1.0`, see [rack/rack@c859bbf](https://github.com/rack/rack/commit/c859bbf7b53cb59df1837612a8c330dfb4147392).
> 
> This is because rails' `CookieStore` extends `Rack::Session::Abstract::Persisted`, see
> 
> https://github.com/rails/rails/blob/f33d52c95217212cbacc8d5e44b5a8e3cdc6f5b3/actionpack/lib/action_dispatch/middleware/session/cookie_store.rb#L40-L42
> 
> .

**Pull request closed without merging**
